### PR TITLE
istioctl:  add a service analyzer for inconsistencies across multiple clusters

### PR DESCRIPTION
**Pull request by hanxiaop**
**Please provide a description of this PR:**
Related to https://github.com/istio/istio/issues/43972 and https://github.com/istio/istio/issues/24345

This PR adds a serviceanalyzer to analyze services across clusters.
The result will be:
```
 istioctl analyze       
Warning [IST0107] (Deployment default/httpbin) Misplaced annotation: istio.io/disconnectedAt can only be applied to WorkloadEntry
Warning [IST0140] (VirtualService default/bookinfo) Subset in virtual service default/productpage has no effect on ingress gateway default/bookinfo requests
Warning [IST0170] (Service default/helloworld-inconsistent) The service helloworld-inconsistent in namespace "default" is inconsistent across clusters ["work-cluster2-130" "work-cluster1-20"], which can lead to undefined behaviors. The inconsistent behaviors are: service ports are inconsistent.
...
```

The approach is using the remote secret that has some accesses to the remote kube api-server, to read and analyze the config.

This PR also enables the analysis framework to recognize resources at the cluster scope, then we can fix old single cluster analyzers or add new analyzers for more multicluster configs.

### CODE REVIEW

File: istioctl/cmd/analyze.go
Comment (john-a-joyce) [inline istioctl/cmd/analyze.go:231]: Will this len be 0 in the single cluster case?  It seemed like that might be the case from line 154 of istiod_analyze.go.   I didn't fully analyze since it's a Draft.

File: istioctl/cmd/analyze.go
Comment (howardjohn) [inline istioctl/cmd/analyze.go:313]: Is this the first istioctl command that accepts multiple clusters at once?

File: pkg/config/analysis/msg/messages.yaml
Comment (howardjohn) [inline pkg/config/analysis/msg/messages.yaml:651]: It would be helpful to indicate in the message why its inconsistent

File: istioctl/cmd/analyze.go
Comment (hanxiaop) [reply istioctl/cmd/analyze.go:231]: Yes, it will maintain the same behavior as before if it's doing single cluster analysis. Only specifying multicluster contexts flag will lead to `MultiClusterResults` being greater than 0.

File: istioctl/cmd/analyze.go
Comment (hanxiaop) [reply istioctl/cmd/analyze.go:313]: Yes, this is the first one. May need to polish the flag name and the way to accept config files/contexts.

File: pkg/config/analysis/msg/messages.yaml
Comment (hanxiaop) [reply pkg/config/analysis/msg/messages.yaml:651]: Thanks, I'll revise with more details.

**Comment by hanxiaop** [comment]
The new results will be:
```
 istioctl analyze -n multicluster -M cluster1 -M cluster2
Analysis results for cluster cluster1:
Info [IST0102] (Namespace multicluster) The namespace is not enabled for Istio injection. Run 'kubectl label namespace multicluster istio-injection=enabled' to enable it, or 'kubectl label namespace multicluster istio-injection=disabled' to explicitly mark it as not needing injection.

Analysis results for cluster cluster2:
Info [IST0102] (Namespace multicluster) The namespace is not enabled for Istio injection. Run 'kubectl label namespace multicluster istio-injection=enabled' to enable it, or 'kubectl label namespace multicluster istio-injection=disabled' to explicitly mark it as not needing injection.

Analysis results based on multi-clusters:
Warning [IST0161] (Service multicluster/helloworld) The service helloworld in namespace "multicluster" is inconsistent across clusters ["cluster1" "cluster2"], which can lead to undefined behaviors. The inconsistent behaviors are: service ports are inconsistent.
Warning [IST0161] (Service multicluster/sleep) The service sleep in namespace "multicluster" is inconsistent across clusters ["cluster1" "cluster2"], which can lead to undefined behaviors. The inconsistent behaviors are: service has mixed mode like headless and clusterIP; service ports are inconsistent.

```

**Comment by hanxiaop** [comment]
Kindly ping @istio/wg-user-experience-maintainers, anyone has any thoughts on this feature?

**Comment by hanxiaop** [comment]
Link the related issue https://github.com/istio/istio/issues/24345

**Comment by therealmitchconnors** [comment]
We have multiple types of multicluster architectures.  Which modes does this command support?  Is there a supported mode where config is spread across multiple clusters, which would make analysis useful?

**Comment by hanxiaop** [comment]
> We have multiple types of multicluster architectures. Which modes does this command support? Is there a supported mode where config is spread across multiple clusters, which would make analysis useful?

@therealmitchconnors This isn't strictly limited to specific deployment models, but it helps differentiate configurations from different clusters, as seen in the [origin](https://github.com/istio/istio/pull/44123/files#diff-bc9d63daadcde7c08f1f921729d8936e3c55ecf47463e28e4f9a5196a88f32c0R30-R31).

In a primary-remote model, for example, we can check the consistency of services across different clusters. In a multi-primary model, we can verify if policies distributed across multiple clusters are consistent, helping to avoid unintended behaviors, among other things.

The effectiveness of this approach will depend on how the developer designs the analyzer, as all resources from the clusters, along with their locations, can be obtained in the multicluster analysis context.

**Comment by hanxiaop** [comment]
@howardjohn @therealmitchconnors Do you have any ideas on this? It's useful for our multicluster deployments and could serve as a starting point for multicluster analysis. One thing I'm uncertain about is whether accepting multiple contexts here is good or should we do something else, since there's currently no cmd doing anything like this. Or should I bring this to the WG meeting for more feedback?

**Comment by hanxiaop** [comment]
> @howardjohn @therealmitchconnors Do you have any ideas on this? It's useful for our multicluster deployments and could serve as a starting point for multicluster analysis. One thing I'm uncertain about is whether accepting multiple contexts here is good or should we do something else, since there's currently no cmd doing anything like this. Or should I bring this to the WG meeting for more feedback?

Another option may be connect to Istiod and fetch data?

**Comment by hanxiaop** [comment]
So basically the idea is:
when users specify the `--multi-cluster` option in `istioctl analyze`, the remote secrets that are created and applied in this cluster will also be fetched. This provides configurations for istioctl to carry out a multi-cluster analysis. This approach works if the remote api-servers are available to be connected after the multi-cluster deployment setup, regardless of the deployment model. 
However, there is one concern: the kubeconfig in the remote secret may use the `istiod-reader-service-account`, which may not have sufficient permissions to list all the configurations needed for the analysis.

Any thoughts on this approach? @howardjohn @therealmitchconnors @stevenctl

**Comment by stevenctl** [comment]
I don't think istio-reader will have sufficient permission to do

> we can verify if policies distributed across multiple clusters are consistent

It will end up being mostly limited to Service/Endpoints.

**Comment by hanxiaop** [comment]
> > we can verify if policies distributed across multiple clusters are consistent
> 
> I don't think istio-reader will have sufficient permission to do this. It will end up being mostly limited to Service/Endpoints.

@stevenctl  Sorry for any confusion. That comment was made when I was using the old approach, which allowed users to specify the config and context with new flags and granted them full permissions. However, none of istioctl command did this before, and it added much complexity of using the multicluster analysis feature.

I have now revised this to the remote secret approach, which seems better in my opinion. However, as you've mentioned, which is my concern as well, the resource permissions will likely be limited for security reasons. Do you have any suggestions?

**Comment by hanxiaop** [comment]
not stale, needs to find a better way to obtain the multicluster resources

**Comment by hanxiaop** [comment]
So essentially, the reader service account was missing a few 'list' permissions for the analysis, primarily for deployments and webhooks. These 'watch' permissions have now been added to the reader service account, and the multi-cluster analysis is working as expected.

**Comment by hanxiaop** [comment]
Like:
```
istioctl analyze --multi-cluster   
Error [IST0101] (Gateway default/bookinfo-gateway) Referenced selector not found: "istio=ingressgateway"
Warning [IST0158] (Namespace default) The Istio proxy images of the pods running in the namespace do not match the image defined in the injection configuration (pod names: [ratings-v1-6484c4d9bb-p8jm8 reviews-v1-5b5d6494f4-n9x6z reviews-v3-5b9bd44f4-fjl9p reviews-v2-5b667bcbf8-nqrkj productpage-v1-675fc69cf-j6nvt]). This often happens after upgrading the Istio control-plane and can be fixed by redeploying the pods.
Warning [IST0163] (Service default/httpbin) The service httpbin in namespace "default" is inconsistent across clusters ["default" "cluster-13900"], which can lead to undefined behaviors. The inconsistent behaviors are: service ports are inconsistent.
Error: Analyzers found issues when analyzing namespace: default.
```

**Review by dhawton — APPROVED** [review]
RN looks good.

**Comment by hanxiaop** [comment]
@therealmitchconnors @howardjohn @stevenctl PTAL, thanks. See the integration test for the example.

File: pkg/config/analysis/msg/messages.yaml
Comment (linsun) [inline pkg/config/analysis/msg/messages.yaml:608]:
```suggestion
    description: "The services live in different clusters under multi-cluster deployment model are inconsistent"
```

Comment (linsun) [reply pkg/config/analysis/msg/messages.yaml:608]: would be good to clarify how inconsistent is defined -- via service name or its namespace?

File: manifests/charts/istio-control/istio-discovery/templates/reader-clusterrole.yaml
Comment (linsun) [inline manifests/charts/istio-control/istio-discovery/templates/reader-clusterrole.yaml:42]: Why does the analyzer require more roles? adding more roles requires justification for many companies.

**Comment by linsun** [comment]
Could you explain why this is needed? It is not clear if we have users asking for it.  Thanks!

File: pkg/config/analysis/msg/messages.yaml
Comment (hanxiaop) [reply pkg/config/analysis/msg/messages.yaml]: Thanks

File: manifests/charts/istio-control/istio-discovery/templates/reader-clusterrole.yaml
Comment (hanxiaop) [reply manifests/charts/istio-control/istio-discovery/templates/reader-clusterrole.yaml:42]: The multi-cluster analyze actually does not require deployment, as its pods already exist in the specified in-cluster format. As for configmap, we only need the injector configmap. However, the analyzers rely on the CRD watcher, which necessitates watch permissions; otherwise, the analysis will not function properly.

Comment (hanxiaop) [reply manifests/charts/istio-control/istio-discovery/templates/reader-clusterrole.yaml:42]: Our Istiod clusterrole is basically a superset of the permissions granted to the reader service account. The reader service account seems to be out-of-date and even lacks some of our API groups. I was thinking adding only the minimal read permissions necessary for analyze to minimize security risks (This will only run in istiod and istioctl), since we cannot using the istiod clusterrole for multi-cluster setup - it's too risky.

**Comment by hanxiaop** [comment]
> Could you explain why this is needed? It is not clear if we have users asking for it. Thanks!

@linsun We have https://github.com/istio/istio/issues/43972 and https://github.com/istio/istio/issues/24345. We've also encountered cases where customers have incorrectly configured their services, leading to significant time spent on debugging what turned out to be simple static configuration issues. While we did have a multi-cluster analyzer in the past, it was designed for analyzing control plane setups and has not yet been refactored ([Issue 28906](https://github.com/istio/istio/issues/28906)). With this new analyzer, we can evaluate various scenarios related to multi-cluster configurations, including both Istio configs and user applications.

File: manifests/charts/istio-operator/templates/clusterrole.yaml
Comment (linsun) [inline manifests/charts/istio-operator/templates/clusterrole.yaml:49]: regarding these permissions added, is it possible to make them optional?  I ask because we would not want to use this.

**Comment by linsun** [comment]
> > Could you explain why this is needed? It is not clear if we have users asking for it. Thanks!
> 
> @linsun We have #43972 and #24345. We've also encountered cases where customers have incorrectly configured their services, leading to significant time spent on debugging what turned out to be simple static configuration issues. While we did have a multi-cluster analyzer in the past, it was designed for analyzing control plane setups and has not yet been refactored #28906. With this new analyzer, we can evaluate various scenarios related to multi-cluster configurations, including both Istio configs and user applications.

Thanks for the background info @hanxiaop! Is it possible to make this function and all the permissions opt-in as not everyone needs it.

**Comment by hanxiaop** [comment]
> > > Could you explain why this is needed? It is not clear if we have users asking for it. Thanks!
> > 
> > 
> > @linsun We have #43972 and #24345. We've also encountered cases where customers have incorrectly configured their services, leading to significant time spent on debugging what turned out to be simple static configuration issues. While we did have a multi-cluster analyzer in the past, it was designed for analyzing control plane setups and has not yet been refactored #28906. With this new analyzer, we can evaluate various scenarios related to multi-cluster configurations, including both Istio configs and user applications.
> 
> Thanks for the background info @hanxiaop! Is it possible to make this function and all the permissions opt-in as not everyone needs it.

Thanks for the great suggestion, I'll revise it.

File: manifests/charts/istio-operator/templates/clusterrole.yaml
Comment (hanxiaop) [reply manifests/charts/istio-operator/templates/clusterrole.yaml:49]: @linsun I've removed permissions related to deploy/configmap/webhooks. However I believe it makes sense to include telemetry and extensions permissions in our reader clusterroles. Last time the update to the file was in July  2021, and since the APIs were added after that date, it seems to be outdated - although there was no issue reported since it's not used in the remote cluster service discovery.

**Comment by hanxiaop** [comment]
not stale, will try to optimize this PR again

**Comment by hanxiaop** [comment]
@therealmitchconnors @howardjohn @linsun @istio/wg-user-experience-maintainers @istio/wg-networking-maintainers Ping some guys to take a look. This feature could be beneficial for users with multicluster setups https://istio.io/latest/docs/setup/install/multicluster/. Remote secrets are used to gather resources in the remote cluster. The reader service account already grants sufficient permissions for today's analyzers, although it lacks some permissions for certain Istio API groups. Therefore, I have added them.

**Review by hzxuzhonghu — COMMENTED** [review]
integration test looks good

File: pkg/kube/client.go
Comment (hanxiaop) [reply pkg/kube/client.go]: This is not needed, thanks.

**Review by hzxuzhonghu — APPROVED** [review]

**Review by kebe7jun — APPROVED** [review]

**Comment by hanxiaop** [comment]
@istio/wg-user-experience-maintainers @istio/wg-environments-maintainers PTAL, thanks

**Comment by zirain** [comment]
only ServiceAnalyzer support multicluster?

**Comment by hanxiaop** [comment]
> only ServiceAnalyzer support multicluster?

@zirain No, this is just the beginning of support that unblocks the multicluster feature. Later on, anyone will be able to write different analyzers as they will know which cluster the resource resides in.

**Comment by zirain** [comment]
> > only ServiceAnalyzer support multicluster?
> 
> @zirain No, this is just the beginning of support that unblocks the multicluster feature. Later on, anyone will be able to write different analyzers as they will know which cluster the resource resides in.

so I think the pr title is not exact right?

**Comment by hanxiaop** [comment]
> > > only ServiceAnalyzer support multicluster?
> > 
> > 
> > @zirain No, this is just the beginning of support that unblocks the multicluster feature. Later on, anyone will be able to write different analyzers as they will know which cluster the resource resides in.
> 
> so I think the pr title is not exact right?

@zirain I've updated the description with two related issues

> Related to https://github.com/istio/istio/issues/43972 and https://github.com/istio/istio/issues/24345

Maybe a "partial support for multicluster" is more accurate, but it's still hard to name the title with the exact thing this PR is doing. I intend to not include more things, like new analyzers in this PR, because it's already very large. We can gradually add more features after this one is merged.
BTW the ServiceAnalyzer is just a new one for the multicluster services. If you are currently following the multicluster setup, the configurations in the remote cluster will also be analyzed by the old analyzers.

**Comment by hanxiaop** [comment]
@zirain Adjusted the title for better clarity. PTAL, Thanks!

**Review by zirain — APPROVED** [review]

**Pull request merged**
